clientconn_test.go

```diff
@@ -64,72 +64,60 @@ func assertState(wantState connectivity.State, cc *ClientConn) (connectivity.Sta
 
 func TestDialWithMultipleBackendsNotSendingServerPreface(t *testing.T) {
 	defer leakcheck.Check(t)
-	numServers := 2
-	servers := make([]net.Listener, numServers)
-	var err error
-	for i := 0; i < numServers; i++ {
-		servers[i], err = net.Listen("tcp", "localhost:0")
+	lis1, err := net.Listen("tcp", "localhost:0")
+	if err != nil {
+		t.Fatalf("Error while listening. Err: %v", err)
+	}
+	defer lis1.Close()
+	lis1Addr := resolver.Address{Addr: lis1.Addr().String()}
+	lis1Done := make(chan struct{})
+	// 1st listener accepts the connection and immediately closes it.
+	go func() {
+		defer close(lis1Done)
+		conn, err := lis1.Accept()
 		if err != nil {
-			t.Fatalf("Error while listening. Err: %v", err)
+			t.Errorf("Error while accepting. Err: %v", err)
+			return
 		}
+		defer conn.Close()
+	}()
+
+	lis2, err := net.Listen("tcp", "localhost:0")
+	if err != nil {
+		t.Fatalf("Error while listening. Err: %v", err)
 	}
-	dones := make([]chan struct{}, numServers)
-	for i := 0; i < numServers; i++ {
-		dones[i] = make(chan struct{})
-	}
-	for i := 0; i < numServers; i++ {
-		go func(i int) {
-			defer func() {
-				close(dones[i])
-			}()
-			conn, err := servers[i].Accept()
-			if err != nil {
-				t.Errorf("Error while accepting. Err: %v", err)
-				return
-			}
-			defer conn.Close()
-			switch i {
-			case 0: // 1st server accepts the connection and immediately closes it.
-			case 1: // 2nd server accepts the connection and sends settings frames.
-				framer := http2.NewFramer(conn, conn)
-				if err := framer.WriteSettings(http2.Setting{}); err != nil {
-					t.Errorf("Error while writing settings frame. %v", err)
-					return
-				}
-				conn.SetDeadline(time.Now().Add(time.Second))
-				buf := make([]byte, 1024)
-				for { // Make sure the connection stays healthy.
-					_, err = conn.Read(buf)
-					if err == nil {
-						continue
-					}
-					if nerr, ok := err.(net.Error); !ok || !nerr.Timeout() {
-						t.Errorf("Server expected the conn.Read(_) to timeout instead got error: %v", err)
-					}
-					return
-				}
-			}
-		}(i)
-	}
+	defer lis2.Close()
+	lis2Done := make(chan struct{})
+	lis2Addr := resolver.Address{Addr: lis2.Addr().String()}
+	// 2nd listener should get a connection attempt since the first one failed.
+	go func() {
+		defer close(lis2Done)
+		_, err := lis2.Accept()
```

Comment:
We are not closing this conn; should we?

Reply:
No, that's intentional. It prevents the client from shutting down too quickly. When the listener shuts down it will terminate the conn, so it's OK to not close individual conns.

Reply:
Sorry, it's the client.Close that closes the conn, not the lis.Close. And, now that I look at this again, I don't think there's a problem with the client side shutting down too quickly (I had thought this test checks more than it does, but it really only checks that the connection was established o.O).

Comment added!